src/tools/crossgen2/Common/TypeSystem/Common/MethodDesc.ToString.cs

```diff
@@ -49,7 +49,18 @@ public override string ToString()
             // (Skipping return type to keep things short)
             sb.Append(OwningType);
             sb.Append('.');
-            sb.Append(Name);
+            try
+            {
+                sb.Append(Name);
```

Comment:
My thinking was along the lines of adding `DiagnosticName` property for methods and `DiagnosticName`/`DiagnosticNamespace` for DefTypes that would always succeed - it would either return the name in metadata, or the token. That way we wouldn't need so many try/catches around grabbing the name or `FormatOptions.UseDiagnosticName` - the formatter would always go for the diagnostic name that has a usable thing and is never allowed to throw (except maybe OOM).

The `.ToString.cs` dot files can depend on the newly added `.Diagnostic.cs` dot files (I can't come up with a scenario where we would not want that) - we shouldn't need the partial methods that indirect into it. Or is there a situation that I'm missing?

Reply:
I buy that the .ToString.cs files can depend on .Diagnostic.cs files as well as the DebugNameFormatter, but I don't think it makes sense to refactor the try's/catches into the leaf nodes or to provide a DiagnosticNamespace property.

1. A DiagnosticNamespace doesn't make sense. The various patterns that are able to produce failsafe names don't consistently have a means for expressing a pair of names for the type.
2. Moving a try/catch into the EcmaType doesn't reduce the need for try/catch at the DebugNameFormatter level. That type (as of this change) is designed to attempt to be resilient to certain forms of incorrectly operating logic including cases where the developer of the typesystem was wrong about the reliability of the code they attempted to write into the DiagnosticName apis. In general, for maximally resilient logic, one should place the exception handling at the outermost possible place, not the innermost. Unfortunately, to make the printing pretty, even in the presence of exceptions, that implies that there are a variety of catches in the type name processing logic. I see this as correct and providing the best approach.

Reply:
Also, I am happy to remove the FormatOptions for diagnostic names, it seems a bit overkill, looking at the change as a whole.

Reply:
I was looking for a way to avoid special casing in the formatter that is never tested. It's easier to review and find issues in a try/catch at the leaf nodes than in a nest of try catches.

> A DiagnosticNamespace doesn't make sense.

In the runtime type system flavor for this, we already have a precedent for a diagnostic name and namespace. The namespace doesn't make much sense when there's no information, so it's just empty in that case; the formatted result is same. I think a DiagnosticName/Namespace that always returns _something_ (either the real thing, or a fallback) results in an easier to use API.

> Moving a try/catch into the EcmaType doesn't reduce the need for try/catch at the DebugNameFormatter level. That type (as of this change) is designed to attempt to be resilient to certain forms of incorrectly operating logic including cases where the developer of the typesystem was wrong about the reliability of the code they attempted to write into the DiagnosticName apis.

Type system developer needs to follow various rules (immutability, thread safety, etc.). The way I look at it, diagnostic name never throwing is just another of those invariants (we could add comments in that sense). Things are pretty bad if we can't trust the type system implementor and we can't protect from that in general.

> Also, I am happy to remove the FormatOptions for diagnostic names, it seems a bit overkill, looking at the change as a whole.

Yes please!

Reply:
I disagree with your approach here, but I don't think its criticallly important. In general, I would rather see incorrect results happen in these cases than failures to catch, but I can see that there is value in the other approach. As I don't consider this critical, I've implemented the DiagnosticNamespace and reduced the catches in the DebugTypeNameFormatter.